### ISSUE TITLE
Fortran artifacts created with GCC 10 may not work on `x86_64-w64-mingw32`

### ISSUE THREAD

**Issue by sloede** [JuliaPackaging/BinaryBuilder.jl#1055]
### TL;DR
`libgfortran-5` shipped with Julia should maybe be built with GCC 10 on all relevant platforms.

### Long(ish) version
Over the past days I tried to get an executable binary to work on `x86_64-w64-mingw32` that I had created sucessfully with `BinaryBuilder.jl` for a JLL package. It is a pure Fortran tool with no external library dependencies. In `build_tarballs.jl`, I used
```julia
platforms = expand_gfortran_versions(platforms)
```
and (at the recommendation of the authors of the tool in question)
```julia
preferred_gcc_version = v"10.2.0"
```
The creation went smoothly, but when I tried to execute the binary from the JLL package on a Windows platform, it resulted in a "silent" error, i.e., no error message but an exit code of `3221225785`. The full output can be seen, e.g., in this [GitHub Actions log](https://github.com/trixi-framework/HOHQMesh.jl/runs/2513227421?check_suite_focus=true#step:6:53), but it was reproducible on other Windows platforms as well.

Searching for the exit code did not help, except that I found hints that it could be a problem with missing/wrong DLLs (shared libraries). The Julia binary directory with `julia` and the appropriate `libgfortran-5.dll` files was on the `PATH`, so this couldn't be an issue either.

I then tried to use the tool [Dependencies](https://github.com/lucasg/Dependencies) to figure out which DLLs were loaded (or not). This yielded the information, that there was indeed an issue with `libgfortran-5.dll`:
![image (1)](https://user-images.githubusercontent.com/3637659/117494273-645fd000-af74-11eb-9dc1-0903d9cfb5d0.png)

However, trying to further investigate `libgfortran-5.dll` by loading it directly in `Dependencies` did not bear fruit, since the error seemed to have disappeared:
![image (2)](https://user-images.githubusercontent.com/3637659/117494401-8d806080-af74-11eb-86d7-894bf8eab112.png)

So far I had tried to not just execute the problematic artifact through Julia, but also directly via the Windows PowerShell, which also silently failed. I then tried to run it in `cmd.exe`, and aha! There was the first somewhat useful error message,
![image (3)](https://user-images.githubusercontent.com/3637659/117494619-db956400-af74-11eb-8319-a7cac984c6f6.png)

which roughly translates to _""function entry point `_gfortran_os_error_at` was not found in DLL"_. A first real clue!

Armed with this new knowledge, I went back to the `Dependencies` tool and tried to see if I could find something related to that and, indeed, the symbol `_gfortran_os_error_at` was the only one marked in red (whatever that exactly means):
![image (4)](https://user-images.githubusercontent.com/3637659/117495107-8017a600-af75-11eb-8efc-2fd1ed0509bc.png)

So, I reapplied some Google-fu (OK, it wasn't hard by now, it was like the third result) and I came across this [helpful writeup](https://www.mjr19.org.uk/sw/gfortran.html) (reproduced here in case the original source gets removed):
> Even though gfortran versions 8 to 10 share the same major version of their library, sometimes a gfortran 10 compiled program will work with a gfortran 8 library, and sometimes it will not. [...], some simple codes compiled with gfortran 10 make no use of run-time library features more recent than gfortran 8, and will run with an older `libgfortran.so.5`. [...] So here the problem is the internal function **`_gfortran_os_error_at`**.

Further mad googling revealed that indeed, `os_error_at` had been added in GCC 10 (see [commit](https://gcc.gnu.org/legacy-ml/gcc-cvs/2019-08/msg00643.html)).

But wait a second... didn't I explicitly request GCC 10 in my `build_tarballs.jl`?

Thus, after a long journey over multiple days, down the rabbit hole of Windows dll debugging and back, the simple fix that made everything whole again was to downgrade the preferred GCC version in `build_tarballs.jl` to `v9.1.0`. That's it. Now everything works perfectly.

Sorry for the long issue, but this is catharsis for me, after almost having given up on getting this to work on Windows at all. 

### Possible solution
To me, it seems like this might be fixable by using a newer GCC version to build the `libgfortran-5.{dll,so,dylib}` shared libraries that are shipped with Julia. However, I am no expert on this and I don't know if this might not create new issues (e.g., if using this library from a binary that was built with GCC <10). Also, I don't know why it's only an issue on Windows and not on any other platform.#1050 

P.S.: Thanks @giordano for bearing with me over the past few days on Slack!

**Comment by giordano** [comment]
Can you please try to replace your Julia's libgfortran-5.dll with the one in https://github.com/JuliaBinaryWrappers/CompilerSupportLibraries_jll.jl/releases/download/CompilerSupportLibraries-v0.4.2%2B1/CompilerSupportLibraries.v0.4.2.x86_64-w64-mingw32-libgfortran5.tar.gz?

**Comment by giordano** [comment]
Or, alternatively, try with Julia [nightly](https://julialang.org/downloads/nightlies/) which already ships that version of libgfortran

**Comment by sloede** [comment]
I tried the following: I downloaded the current Julia nightlies/portable (julia-6726ae89f0), put the bin/ as the only Julia installation on the PATH, and tried to run the GCC10-built artifact again - same issue as before. I also viewed the artifact in `Dependencies` - same issue about missing `_gfortran_os_error_at´:
![image](https://user-images.githubusercontent.com/3637659/117508818-7ac45680-af89-11eb-9ad1-08726a9d7c4b.png)
![image](https://user-images.githubusercontent.com/3637659/117508841-81eb6480-af89-11eb-8e3d-25835388e9c8.png)


I then downloaded the CSL you linked to [above](https://github.com/JuliaBinaryWrappers/CompilerSupportLibraries_jll.jl/releases/download/CompilerSupportLibraries-v0.4.2%2B1/CompilerSupportLibraries.v0.4.2.x86_64-w64-mingw32-libgfortran5.tar.gz) and replaced the `libgfortran-5.dll` in Julia's bin/ directory with the one from the downloaded CSL archive. Again, all issues remain as before...

Thus, either I did something wrong, the Julia nightlies for Windows are *not* built with GCC10, or my assumptions about the root cause of this issue are flawed.

**Comment by giordano** [comment]
I suspect `CompilerSupportLibraries` is shipping libraries from GCC 8 (the oldest) instead of 10 (the most recent) for libgfortran 5, which is quite the opposite of what it's supposed to do

**Comment by giordano** [comment]
See https://github.com/JuliaPackaging/Yggdrasil/pull/2946.

**Comment by giordano** [comment]
With the libgfortran from the tarball for `x86_64-w64-mingw32-libgfortran5` in that pull request I get
```console
sandbox:${WORKSPACE} # nm libgfortran-5.dll | grep _gfortran_os_error_at
00000000656ff2dc p .pdata$_gfortran_os_error_at
00000000654f8f70 t .text$_gfortran_os_error_at
0000000065706270 r .xdata$_gfortran_os_error_at
00000000654f8f70 T _gfortran_os_error_at
```
so the symbol is there, contrary to the currently shipped libgfortran5 (because it was from GCC 8).  It'd be great if you could test with this libgfortran.

As a side note, there is a reason why we always build with the _**oldest**_ version of the compiler possible :slightly_smiling_face:

**Comment by sloede** [comment]
🎉 🥳 

With the CSL from https://github.com/JuliaPackaging/Yggdrasil/pull/2946, I get
![image](https://user-images.githubusercontent.com/3637659/117525845-f262a780-afc2-11eb-984f-d99a97ddf4d3.png)
![image](https://user-images.githubusercontent.com/3637659/117526057-ff7f9680-afc2-11eb-8492-d9f0abbe013b.png)
![image](https://user-images.githubusercontent.com/3637659/117526085-10c8a300-afc3-11eb-946b-efe558941b50.png)

Thus
1. No obvious errors shown in `Dependencies` anymore
2. `Dependencies` has not flagged `_gfortran_os_error_at` as red anymore
3. Running the same artifact that failed before on Windows no produces the expected output (i.e., that no help is available yet 😅 )

Thus, the change in the CSL `build_tarballs.jl` does indeed produce a `libgfortran-5.dll` that fixes the originally reported issue 👍

**Comment by sloede** [comment]
With JuliaLang/julia#40807 merged, does this mean that this issue might be resolved with the next Julia patch release?

**Comment by giordano** [comment]
> this issue might be resolved with the next Julia patch release?

Minor release (v1.7), I don't think this will be backported to the v1.6 branch
